File: go/leetcode/155.最小栈.go
/*
 * @lc app=leetcode.cn id=155 lang=golang
 *
 * [155] 最小栈
 *
 * https://leetcode-cn.com/problems/min-stack/description/
 *
 * algorithms
 * Easy (49.91%)
 * Likes:    285
 * Dislikes: 0
 * Total Accepted:    45.8K
 * Total Submissions: 91.3K
 * Testcase Example:  '["MinStack","push","push","push","getMin","pop","top","getMin"]\n[[],[-2],[0],[-3],[],[],[],[]]'
 *
 * 设计一个支持 push，pop，top 操作，并能在常数时间内检索到最小元素的栈。
 * 
 * 
 * push(x) -- 将元素 x 推入栈中。
 * pop() -- 删除栈顶的元素。
 * top() -- 获取栈顶元素。
 * getMin() -- 检索栈中的最小元素。
 * 
 * 
 * 示例:
 * 
 * MinStack minStack = new MinStack();
 * minStack.push(-2);
 * minStack.push(0);
 * minStack.push(-3);
 * minStack.getMin();   --> 返回 -3.
 * minStack.pop();
 * minStack.top();      --> 返回 0.
 * minStack.getMin();   --> 返回 -2.
 * 
 * 
 */

// 最小值也是一个栈，每次发现有更小的值就加入栈，出栈的时候检查是否最小值，是的话从min栈里出去
type MinStack struct {
	Data []int
	min []int
}


/** initialize your data structure here. */
func Constructor() MinStack {
    return MinStack{
		Data: []int{},
		min: []int{},
	}
}


func (this *MinStack) Push(x int)  {
	this.Data = append(this.Data, x)  // 2 0 3
	if len(this.min) <= 0 || this.min[len(this.min)-1] >= x {  // NOTE: 注意=也要加进去
		this.min = append(this.min, x) // 2 0
	}
}


func (this *MinStack) Pop()  {
	if len(this.Data) <= 0 {
		return
	}
	v := this.Data[len(this.Data)-1]  // 3 0 2
	this.Data = this.Data[:len(this.Data)-1]
	if len(this.min) > 0 && v <= this.min[len(this.min)-1] {
		this.min = this.min[:len(this.min)-1] // 2 0 
	}
}


func (this *MinStack) Top() int {
	if len(this.Data) <= 0 {
		return 0
	}
	return this.Data[len(this.Data)-1]
}


func (this *MinStack) GetMin() int {
    if len(this.min) <= 0 {
		return 0
	}
	return this.min[len(this.min)-1]
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: cpp/leetcode/155.最小栈.cpp
/*
 * @lc app=leetcode.cn id=155 lang=cpp
 *
 * [155] 最小栈
 *
 * https://leetcode.cn/problems/min-stack/description/
 *
 * algorithms
 * Medium (58.94%)
 * Likes:    1566
 * Dislikes: 0
 * Total Accepted:    477.5K
 * Total Submissions: 810K
 * Testcase Example:
 * '["MinStack","push","push","push","getMin","pop","top","getMin"]\n[[],[-2],[0],[-3],[],[],[],[]]'
 *
 * 设计一个支持 push ，pop ，top 操作，并能在常数时间内检索到最小元素的栈。
 *
 * 实现 MinStack 类:
 *
 *
 * MinStack() 初始化堆栈对象。
 * void push(int val) 将元素val推入堆栈。
 * void pop() 删除堆栈顶部的元素。
 * int top() 获取堆栈顶部的元素。
 * int getMin() 获取堆栈中的最小元素。
 *
 *
 *
 *
 * 示例 1:
 *
 *
 * 输入：
 * ["MinStack","push","push","push","getMin","pop","top","getMin"]
 * [[],[-2],[0],[-3],[],[],[],[]]
 *
 * 输出：
 * [null,null,null,null,-3,null,0,-2]
 *
 * 解释：
 * MinStack minStack = new MinStack();
 * minStack.push(-2);
 * minStack.push(0);
 * minStack.push(-3);
 * minStack.getMin();   --> 返回 -3.
 * minStack.pop();
 * minStack.top();      --> 返回 0.
 * minStack.getMin();   --> 返回 -2.
 *
 *
 *
 *
 * 提示：
 *
 *
 * -2^31 <= val <= 2^31 - 1
 * pop、top 和 getMin 操作总是在 非空栈 上调用
 * push, pop, top, and getMin最多被调用 3 * 10^4 次
 *
 *
 */
#include <stack>

// @lc code=start
class MinStack {
public:
  MinStack() : data_(), min_() {}

  void push(int val) {
    data_.push(val);
    if (min_.empty() || val <= min_.top()) {
      min_.push(val);
    }
  }

  void pop() {
    if (!min_.empty() && min_.top() == data_.top()) {
      min_.pop();
    }
    data_.pop();
  }

  int top() { return data_.top(); }

  int getMin() {
    if (min_.empty()) {
      return -1;
    }
    return min_.top();
  }

private:
  std::stack<int> data_;
  std::stack<int> min_;  // 遇到更小的进栈，出栈判断是否将最小值出栈
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end
